### PR TITLE
Add FarSeg model

### PR DESCRIPTION
**Pull request by Z-Zheng**
- Add `torchgeo.models.FarSeg` from ["Foreground-Aware Relation Network for Geospatial Object Segmentation
    in High Spatial Resolution Remote Sensing Imagery", Zheng et al. (CVPR 2020)](https://arxiv.org/pdf/2011.09766.pdf)

### CODE REVIEW

File: torchgeo/models/__init__.py
Comment (adamjstewart) [inline torchgeo/models/__init__.py:10]: Lets put these in alphabetical order, so FarSeg first.

File: torchgeo/models/farseg.py
Comment (adamjstewart) [inline torchgeo/models/farseg.py:13]: If you run `isort .` it will resort all of these imports for you. This is required to get the unit tests to pass. Just `pip install isort` if you don't already have it installed. Same for black/flake8/mypy/pydocstyle.

File: torchgeo/models/farseg.py
Comment (adamjstewart) [inline torchgeo/models/farseg.py:20]: It might be better to describe what the model actually does instead of the title of the paper, since the link to the paper is already below.

File: torchgeo/models/farseg.py
Comment (adamjstewart) [inline torchgeo/models/farseg.py:30]: You can skip this, arXiv already has an easy way to get the bibtex for a paper and this won't format well in the docs anyway.

File: torchgeo/models/farseg.py
Comment (adamjstewart) [inline torchgeo/models/farseg.py:46]: I don't think this really works unless `in_channels=3`. In the long run we're planning on shipping resnets pre-trained on multispectral satellite imagery, but we don't have that ready yet. Until then it might be better just to skip this, or raise an error if someone requests a pre-trained model when `in_channels != 3`.

File: torchgeo/models/farseg.py
Comment (adamjstewart) [inline torchgeo/models/farseg.py:39]:
```suggestion
        """Initialize a new FarSeg model.
```

File: torchgeo/models/farseg.py
Comment (adamjstewart) [inline torchgeo/models/farseg.py:72]: All of these functions will need docstrings explaining what they do, what their input arguments are, and what their return values are.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Thanks for adding this! I added a few initial comments, there may be more to follow.

**Comment by isaaccorley** [comment]
Hi @Z-Zheng! TorchGeo's checks require that you run isort/black on your scripts as well as fixing any errors after running flake8, pydocstyle, and mypy before we can merge.

**Comment by calebrob6** [comment]
@adamjstewart, can you look at all the attr-defined errors thrown by mypy? There has to be a better way to deal with those then scattering the "# type: ignore[attr-defined]" things everywhere. If there isn't, can you add them?

@Z-Zheng this mostly looks good to me now -- the only thing I would say is to add more descriptive docstrings if possible and say a little more about what this model is doing.

**Comment by calebrob6** [comment]
Nevermind @adamjstewart -- I think we appeased mypy the hacky way :). Thanks @Z-Zheng!

I'd love a little bit more about the model in the main FarSeg docstring, e.g. what it is good at, what it has been used for, why it is different, but am happy with the implementation.

File: torchgeo/models/farseg.py
Comment (adamjstewart) [inline torchgeo/models/farseg.py:197]: These lines don't have any test coverage

File: torchgeo/models/farseg.py
Comment (calebrob6) [reply torchgeo/models/farseg.py:67]: `in_channels` has to be 3 for now because the resnet backbones don't naively have an option to take other sized inputs. (note: we can change this by overwriting model.conv1 as I do in the So2Sat trainer)

Comment (adamjstewart) [reply torchgeo/models/farseg.py:67]: In that case should we not have an `in_channels` option at all if it doesn't do anything?

**Comment by adamjstewart** [comment]
Okay, I added a commit to change a few more things:

- [x] Reformat the FarSeg docstring a bit
- [x] Remove defaults from function docstrings (these already show up in the docs anyway)
- [x] Convert `super()` usage to Python 3 style
- [x] Convert helper modules to private classes so they don't appear in the docs
- [x] Remove unused `scale_aware_proj` parameter from `FSRelation` module

I think this is ready to merge now if everyone approves of the above changes.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
